### PR TITLE
Ensure vector similarity correctly limits inner_hits returned for nested kNN

### LINKED ISSUE #111093: We ignore `similarity` when gathering inner_hits for vectors
### Elasticsearch Version

_any_

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

_any_

### Problem Description

When utilizing nested kNN queries & a required similarity, we correctly filter for the top-level docs and similarity is applied correctly. 

However, during the fetch phase, when we gather and score inner_hits, we are ignoring the configured similarity. It needs to be applied.

To fix this, we will need to update `ExactKnnQueryBuilder` to include a similarity threshold. 

Look for usages `if (ctx.convertToInnerHitsRewriteContext() != null)`, and we will also need to update `KnnScoreDocQueryBuilder` to include the similarity used to create it.

### Steps to Reproduce

 - Index nested vectors
 - search by a similarity that hits the max passage value, but should filter a lesser relevant passage
 - ensure that inner_hits are enabled
 - see how the lesser scoring passage is still included.

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by benwtrent**
For nested kNN we support not only similarity thresholds, but also multi-passage search while retrieving more than one nearest passage. 

However, the inner_hits retrieved for the kNN search would ignore the restricted similarity. Meaning, the inner hits would return all passages, not just the ones within the limited similarity and this is confusing.

closes: https://github.com/elastic/elasticsearch/issues/111093

### CODE REVIEW

**Review by john-wagster — APPROVED** [review]
LGTM

**Review by carlosdelest — APPROVED** [review]

File: docs/changelog/111363.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+pr: 111363
+summary: Ensure vector similarity correctly limits `inner_hits` returned for nested
+  kNN
+area: Vector Search
+type: bug
+issues: []
```
Comment (benwtrent) [inline docs/changelog/111363.yaml:6]:
```suggestion
issues: [111093]
```

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
run elasticsearch-ci/packaging-tests-windows-sample

**Comment by benwtrent** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[](https://github.com/elastic/elasticsearch/pull/111426)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**